GitUpKit/Interface/GIImageDiffView.m

```diff
@@ -0,0 +1,226 @@
+@end
+
+@implementation GIImageDiffView
+- (id)initWithRepository:(GCLiveRepository*)repository {
+  self = [super initWithFrame:CGRectZero];
+  self.repository = repository;
+  _percentage = 0.5;
+  [self setupView];
+  return self;
+}
+
+- (void)setupView {
+  self.wantsLayer = true;
+
+  _oldImageBorderLayer = [[CALayer alloc] init];
+  _currentImageBorderLayer = [[CALayer alloc] init];
+  [self.layer addSublayer:_oldImageBorderLayer];
+  [self.layer addSublayer:_currentImageBorderLayer];
+
+  _transparencyCheckerboardLayer = [[CALayer alloc] init];
+  NSBundle* bundle = NSBundle.gitUpKitBundle;
+  NSImage* patternImage = [bundle imageForResource:@"background_pattern"];
+  _checkerboardColor = [NSColor colorWithPatternImage:patternImage];
+  _transparencyCheckerboardLayer.backgroundColor = _checkerboardColor.CGColor;
+  [self.layer addSublayer:_transparencyCheckerboardLayer];
+
+  _currentImageView = [[NSImageView alloc] init];
+  _oldImageView = [[NSImageView alloc] init];
+  [self addSubview:_currentImageView];
+  [self addSubview:_oldImageView];
+
+  _oldImageMaskLayer = [[CALayer alloc] init];
+  _oldImageMaskLayer.backgroundColor = NSColor.blackColor.CGColor;
+  _oldImageView.wantsLayer = true;
+  _oldImageView.layer.mask = _oldImageMaskLayer;
+
+  _currentImageMaskLayer = [[CALayer alloc] init];
+  _currentImageMaskLayer.backgroundColor = NSColor.blackColor.CGColor;
+  _currentImageView.wantsLayer = true;
+  _currentImageView.layer.mask = _currentImageMaskLayer;
+
+  _dividerView = [[NSView alloc] init];
+  [self addSubview:_dividerView];
+
+  _panGestureRecognizer = [[NSPanGestureRecognizer alloc] initWithTarget:self action:@selector(didMoveSplit:)];
+  _clickGestureRecognizer = [[NSClickGestureRecognizer alloc] initWithTarget:self action:@selector(didMoveSplit:)];
+  [self addGestureRecognizer:_panGestureRecognizer];
+  [self addGestureRecognizer:_clickGestureRecognizer];
+}
+
+- (void)setDelta:(GCDiffDelta*)delta {
+  if (delta != _delta) {
+    _delta = delta;
+    [self updateCurrentImage];
+    [self updateOldImage];
+    self.percentage = 0.5;
+  }
+}
+
+- (void)setPercentage:(CGFloat)percentage {
+  _percentage = percentage;
+  [self setNeedsDisplay:true];
+}
+
+- (void)updateCurrentImage {
+  NSError* error;
+  NSString* newPath;
+  if (_delta.newFile.SHA1 != nil) {
+    newPath = [GILaunchServicesLocator.diffTemporaryDirectoryPath stringByAppendingPathComponent:_delta.newFile.SHA1];
+    NSString* newExtension = _delta.newFile.path.pathExtension;
+    if (newExtension.length) {
+      newPath = [newPath stringByAppendingPathExtension:newExtension];
+    }
+    if (![[NSFileManager defaultManager] fileExistsAtPath:newPath]) {
+      [self.repository exportBlobWithSHA1:_delta.newFile.SHA1 toPath:newPath error:&error];
+    }
+  } else {
+    newPath = [self.repository absolutePathForFile:_delta.canonicalPath];
+  }
+  _currentImageView.image = [[NSImage alloc] initWithContentsOfFile:newPath];
```

Comment:
I'm a bit concerned about loading in the file this way in cases where the asset is rather large. For most cases we're probably not dealing with giant assets, but if you do this is going to grind to a halt. We should probably generate a thumbnail of a max size so we cap the load time. Another optimization we should probably have is to separate out the image loading and size calculations. The table view needs to know size (because we don't use automatic height adjustments yet, unfortunately) and the size calculation should be as fast as possible.

Reply:
Yeah, limiting the size in memory definitely sounds advisable!
I took care of that in 991f88908737795da132a25a4a2fef40e01c0364. For now I simply used a somewhat arbitrary maximum image dimension of 4000 pixels, because it was easy to do and seemed like a reasonable limit.
If you think we should change this value or even make it scale dynamically by screen size or something like that let me know 🙂

Reply:
With regards to separating out image loading and size calculations:
Does that mean you would want to make the cells a static height?

Right now we load the new images on `setDelta` and then query their dimensions during the sizing calculations to make sure they aren't scaled improperly and the aspect ratios make sense.
So the only way I could think of to avoid this reliance on in-memory images would be to set an arbitrary height and then cram the images into that space whenever they are ready 🤔